File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge30.py
""" package com.mouredev.weeklychallenge2022

/*
 * Reto #30
 * MARCO DE PALABRAS
 * Fecha publicación enunciado: 26/07/22
 * Fecha publicación resolución: 01/08/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea una función que reciba un texto y muestre cada palabra en una línea, formando
 * un marco rectangular de asteriscos.
 * - ¿Qué te parece el reto? Se vería así:
 *   **********
 *   * ¿Qué   *
 *   * te     *
 *   * parece *
 *   * el     *
 *   * reto?  *
 *   **********
 *
 * Información adicional:
 * - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
 * - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
 * - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
 *
 */

fun main() {
    drawFrame("¿Qué te parece el reto?")
    drawFrame("¿Qué te     parece el reto?")
    drawFrame("¿Cuántos retos de código de la comunidad has resuelto?")
}

private fun drawFrame(text: String) {

    val words = text.split(" ")
    var maxLength = 0
    words.forEach { word ->
        if (word.length > maxLength) {
            maxLength = word.length
        }
    }

    println("*".repeat(maxLength + 4))

    words.forEach { word ->
        if (word.isNotEmpty()) {
            println("* $word${" ".repeat(maxLength - word.length)} *")
        }
    }

    println("*".repeat(maxLength + 4))
}
 """

def text_frame(text: str):

    lines = text.split()
    max_char = 0

    for line in lines:
        if max_char < len(line):
            max_char = len(line)

    print("*" * (max_char + 4))

    for line in lines:
        print("* " + line + " " * (max_char - len(line)) + " *")

    print("*" * (max_char + 4))


text_frame("¿Qué te parece el reto?")
text_frame("¿Qué te     parece el reto?")
text_frame("¿Cuántos retos de código de la comunidad has resuelto?")

File: app/src/main/java/com/mouredev/weeklychallenge2022/Challenge30.dart
/*
 * Reto #30
 * MARCO DE PALABRAS
 * Fecha publicación enunciado: 26/07/22
 * Fecha publicación resolución: 01/08/22
 * Dificultad: FÁCIL
 *
 * Enunciado: Crea una función que reciba un texto y muestre cada palabra en una línea, formando
 * un marco rectangular de asteriscos.
 * - ¿Qué te parece el reto? Se vería así:
 *   **********
 *   * ¿Qué   *
 *   * te     *
 *   * parece *
 *   * el     *
 *   * reto?  *
 *   **********
 *
 * Información adicional:
 * - Usa el canal de nuestro discord (https://mouredev.com/discord) "🔁reto-semanal" para preguntas, dudas o prestar ayuda a la comunidad.
 * - Puedes hacer un Fork del repo y una Pull Request al repo original para que veamos tu solución aportada.
 * - Revisaré el ejercicio en directo desde Twitch el lunes siguiente al de su publicación.
 * - Subiré una posible solución al ejercicio el lunes siguiente al de su publicación.
 *
 */

enum Align { left, center, right }

// ********** ********** **********
// * ¿Qué   * *  ¿Qué  * *   ¿Qué *
// * te     * *   te   * *     te *
// * parece * * parece * * parece *
// * el     * *   el   * *     el *
// * reto?  * * reto?  * *  reto? *
// ********** ********** **********
void main(List<String> args) {
  String wrapText(String input,
      {String char = '*', String space = ' ', Align align = Align.left}) {
    String _char = char.substring(0, 1);
    String _space = space.substring(0, 1);
    String textWrapped = '';
    List<String> lines = input.split(RegExp(r'\s+'));
    if (lines.isNotEmpty) {
      int width = 0;
      for (var line in lines) width = width > line.length ? width : line.length;

      textWrapped = '${_char * (width + 4)}\n'; //lado superior
      for (var line in lines) {
        textWrapped += (_char + _space); //lado izquierdo
        //contenido
        switch (align) {
          case Align.left:
            textWrapped += '$line${_space * (width - line.length)}';
            break;
          case Align.center:
            int spaceLeft = ((width - line.length) ~/ 2);
            int spaceRight = width - line.length - spaceLeft;
            textWrapped += '${_space * spaceLeft}$line${space * spaceRight}';
            break;
          case Align.right:
            textWrapped += '${_space * (width - line.length)}$line';
            break;
        }
        textWrapped += '${(_space + _char)}\n'; //lado derecho
      }
      textWrapped += '${_char * (width + 4)}\n'; //lado inferior
    }
    return textWrapped;
  }

  print(wrapText('¿Qué te parece el reto?', align: Align.left));
  print(wrapText('¿Qué te parece el reto?', align: Align.center));
  print(wrapText('¿Qué te parece el reto?', align: Align.right));
}
